### DISCUSSION TITLE
Is it possible to use a nested private class?

### DISCUSSION THREAD

**Discussion by glen-84** [SteveDunn/Vogen#371] [Q&A] [answered]
This doesn't seem to work:

```cs
public sealed class Example
{
    // ...

    [ValueObject]
    private partial class Test { }
}
```

**Comment by SteveDunn** [comment]
Not currently, and I wasn't planning on enabling it. What scenario would you like to use this in? I'm happy to add it for any useful situations.

**Reply by glen-84** [reply]
I want to return an instance of a private value object from the parent class, so that the value object cannot be constructed externally (it's returned via an interface). If you have an instance of this value object, then you know that it has been constructed by the parent class (in a specific way).

Hopefully that makes sense, otherwise I can post some example code. 🙂

**Reply by SteveDunn** [reply]
I see. Sounds like a good idea. Some example code would be great!

**Reply by glen-84** [reply]
```cs
public sealed class ContentSanitizer
{
    public static ISanitizedContent GetSanitizedContent(string content)
    {
        return new SanitizedContent()
        {
            Value = content.Replace("bad", "good", StringComparison.Ordinal) // Sanitize.
        };
    }

    // Value object.
    private sealed class SanitizedContent : ISanitizedContent
    {
        public required string Value { get; init; }
    }
}

public interface ISanitizedContent
{
    public string Value { get; init; }
}

[Fact]
public void Test()
{
    var sanitizedContent = ContentSanitizer.GetSanitizedContent("bad");

    Assert.Equal("good", sanitizedContent.Value);

    // `SanitizedContent` cannot be externally constructed.
}
```

**Comment by SteveDunn** [comment]
I took another look at this. For your scenario to work as intended, the C# compiler would insist that `ContentSantizer` be `partial` to allow the generation of the inner partial class (the Value Object).

I'm guessing this would open up a huge back door to the safety that you want in your santizer implementation, allowing anyone to extend the type with another partial implementation.

I'm torn between disallowing it, which is does currently, or allowing it if all parent types are partial.  The tricky things is that the generator would have to keep track of the parent types and whether they're classes, structs, records, readonly structs/record structs etc. etc., e.g.

```csharp
public partial class Layer1
{
    internal sealed partial class Layer2
    {
        public partial struct Layer3
        {
            public sealed partial record Layer4
            {
                internal partial record struct Layer5
                {
                    [ValueObject]
                    public partial class MyVo
                    {
                    }
                }
            }
        }
    }
}
```

I don't know if the follow suggestion will be of any use to you, but did you know that Vogen allows for '[normalization](https://github.com/SteveDunn/Vogen/wiki/Normalization)'? Taking your example above, it would be something like:

```csharp
[ValueObject<string>]
public partial class MySantizedThing
{
    private static string NormalizeInput(string input) => input.Replace("bad", "good");
}
```

Normalization happens before Validate is called (if there is a Validate method).

**Reply by glen-84** [reply, marked as answer]
I apologize for the large delay in responding (moved apartments).

Someone would be able to add another `ISanitizedContent` implementation, but they wouldn't be able to change `GetSanitizedContent`, as far as I understand.

With that said, the implementation complexity on the Vogen side might not justify the effort. For now we're just returning a simple record.

Regarding normalization, I am aware of it. The example that I provided is just one part of a system that processes content with one or more "processors", which also have dependencies. So, it's likely not a good fit.

Thanks for getting back to me about this, but I think we're good for now. 👍

**Reply by SteveDunn** [reply]
No worries - thanks again for the feedback!
